src/asm/stages/stage_3.asm

```diff
@@ -0,0 +1,199 @@
+	mov [_pmlt4], eax	# map the first entry in the page map level 4 table to the adress
+						# of the page table directory pointer with rw and present attributes
+	lea eax, [_pdt]
+	or eax, 0x3
+	mov [_pdpt], eax
+
+	mov ecx, 0
+map_pdt_table:
+	mov eax, 0x200000	# 2mb
+	mul ecx 			# multiply eax by the current offset
+	or eax, 0b10000011  # mark the entry with present + readwrite + huge
+	mov [_pdt + ecx * 8], eax # sizeof _pdt entry is 8 and there are 512 entries
+
+	inc ecx
+	cmp ecx, 512
+	jne map_pdt_table
+
+	pop ecx
+	ret
+
+
+enable_paging:
+	# This function enables paging by writing the address of _pmlt4 to cr3
+	# setting the pae flag in cr4
+	# setting the long mode bit in EFER
+	# and setting the paging bit in cr0
+    wbinvd # write back cache
+    mfence # add memory fence
+
+	push ecx
+
+	lea eax, [_pmlt4]
+	mov cr3, eax
+
+	mov eax, cr4
+	or eax, 1 << 5 	# Enable the PAE flag in cr4 (The physical address extesion)
+	mov cr4, eax
+
+	mov ecx, 0xc0000080 # Set the long mode bit in EFER MSR (Extended Feature Enable Register)
+	rdmsr 				# Read model specific register specified in ecx into edx:eax
+	or eax, 1 << 8		# Enable long mode bit
+	wrmsr 				# Write into model specific register
+
+	mov eax, cr0	 	# Enable paging 
+	or eax, 1 << 31
+	mov cr0, eax
+
+	pop ecx
+	ret
+
+.align 4
+zero_idt:
+    .word 0
+    .byte 0
+
+gdt_64:
+	.word 0                    		# Limit (low).
+    .word 0                         # Base (low).
+    .byte 0                         # Base (middle)
+    .byte 0                         # Access.
+    .byte 0                         # Granularity.
+    .byte 0
+codeseg64:
+	.word 0                         # Limit (low).
+    .word 0                         # Base (low).
+    .byte 0                         # Base (middle)
+    .byte 0b10011000                # Access (exec/read).
+    .byte 0b00100000                # Granularity, 64 bits flag, limit19:16.
+    .byte 0                         # Base (high).
+dataseg64:
+    .word 0                         # Limit (low).
+    .word 0                         # Base (low).
+    .byte 0                         # Base (middle)
+    .byte 0b10010000                # Access (read/write).
+    .byte 0b00000000                # Granularity.
+    .byte 0                         # Base (high).
+
+.align 4
+    .word 0                              # Padding to make the "address of the GDT" field aligned on a 4-byte boundary
+
+gdt_64_pointer:
+    .word gdt_64_pointer - gdt_64 - 1    # 16-bit Size (Limit) of GDT.
+    .long gdt_64                            # 32-bit Base Address of GDT. (CPU will zero extend to 64-bit)
+
+.include "src/asm/utils/32_print_string.asm"
+
+cpuid_not_supported_str:		.asciz	"cpuid is not supported, no way to check long mode"
+long_mode_not_supported_str:	.asciz	"long mode is not supported!"
+paging_enabled_str:				.asciz  "Paging is officaly enabled!"
+
+.code64
+long_mode_start:
+	mov rax, 0x2f592f412f4b2f4f
```

Comment:
what the fuck? comment this (printing OKAY or whatever)

Reply:
fixed